src/dmd/expressionsem.d

```diff
@@ -7184,6 +7184,28 @@ private extern (C++) final class ExpressionSemanticVisitor : Visitor
                 return setError();
         }
 
+        static bool isReal(const TY ty) @safe pure nothrow @nogc
+        {
+            return (ty == Tfloat32 ||
+                    ty == Tfloat64 ||
+                    ty == Tfloat80);
+        }
+
+        static bool isImaginary(const TY ty) @safe pure nothrow @nogc
+        {
+            return (ty == Timaginary32 ||
+                    ty == Timaginary64 ||
+                    ty == Timaginary80);
+        }
+
+        if (isReal(exp.to.ty) && isImaginary(exp.e1.type.ty) ||
```

Comment:
Erm...
```suggestion
        if (exp.to.isreal() && exp.e1.type.isimaginary() ||
```

Reply:
No need to reinvent the wheel.

Reply:
I say that, but `isreal` is virtual, right so that's gonna be slower. I'm not sure how picky you are about performance..

Reply:
> Erm...

Done.

Reply:
> Erm...

What about making this warning happen also when converting integer types to imaginary?

Reply:
Then you'd use `exp.to.isimaginary() != exp.e1.type.imaginary()`.